src/NuGet.Core/NuGet.ProjectModel/PublicAPI.Unshipped.txt

```diff
@@ -4,3 +4,4 @@ NuGet.ProjectModel.ProjectRestoreMetadataFrameworkInfo.TargetAlias.get -> string
 NuGet.ProjectModel.ProjectRestoreMetadataFrameworkInfo.TargetAlias.set -> void
 NuGet.ProjectModel.TargetFrameworkInformation.TargetAlias.get -> string
 NuGet.ProjectModel.TargetFrameworkInformation.TargetAlias.set -> void
+static NuGet.ProjectModel.PackageSpecExtensions.ApplyCentralVersionInformation(this NuGet.ProjectModel.PackageSpec project) -> NuGet.ProjectModel.PackageSpec
```

Comment:
The method was moved to the LibraryDependency.

Reply:
I think you misunderstood the point of my comment.  I don't think that CPVM is interesting/important/unique enough to API users to warrant a specific API just for it.  If, for some reason, some enterprise customer is using `PackageSpec` or `LibraryModel` in their app, it's really unfortunate that they have to do:

```cs
var packageSpec = new PackageSpec { ... };
packageSpec.ApplyCPVM();
```

It doesn't matter if it's the `PackageSpec` or `LibraryModel`, or whatever, if they forget to call `ApplyCPVM()`, then the package spec is "broken" and other APIs that use it (like restore) will produce the wrong results. But we can't redesign `PackageSpec` from scratch, as that will be a breaking change, so let's accept that we have to have some method to mutate the object after it's been instantiated.

The problem with `ApplyCPVM()` from an API design point of view is that if some new feature comes in the future and we also need to mutate the `PackageSpec` for that feature, and maybe there's another feature that also requires it, then using this API pattern this PR is proposing, customers (including of NuGet's own code that uses a `PackageSpec`, will have to call:

```cs
var packageSpec = new PackageSpec() { ... };
packageSpec.ApplyCPVM();
packageSpec.ApplyNewFeature();
packageSpec.ApplyAnotherFeature();
```

It's better if there's just 1 method `Prepare()`, or maybe someone can think of a better name, and `Prepare` calls `ApplyCPVM`, `ApplyNewFeature` and `ApplyAnotherFeature`. This way anything that uses the `PackageSpec`, or `LibraryModel`, or whatever, can call just 1 method, and when new versions of NuGet introduce new features, they don't have to change their code even more. Technically, no longer being able to new-up a `new PackageSpec() { ... }` and have it in a "working state" is a breaking change. But we can minimise future breaking changes by having a single `Prepare` public API, and its implementation is calling all the feature-specific `Apply*` that are needed.

Reply:
I moved the logic to version merging to the LibraryDependency because the version changes will apply to the LibraryDependecy types. 
In my opinion a generic Prepare method could be confusing. What would be a good documentation for it? Would a call to the Prepare method be allowed to mutate the inputs received? 

Currently at each entry point (msbuild, VS etc.) the steps to create a package spec are.
1. Instantiate a packageSpec 
2. Create TargetFrameworkInformation and set dependencies, central versions and other properties,
3. Add more properties on the packageSpec. 

The added method allows the callers to create the dependencies with the desired versions before assigning them to the packageSpec.

Reply:
I really wish I added the following comment couple of days ago. I agree with other comments to have a single `public API` that perform all the required mutations on the `package spec` object.

 > 3. Add more properties on the packageSpec.

I was going through various docs to suggest better API for this problem. I found a [`Mutator design pattern`](https://hillside.net/plop/2006/Papers/ACM_Version/The%20Mutator%20Pattern.pdf) to apply a series of successive modifications to a mutable object (`PackageSpec` in this case). 

We can have a static method somewhere that creates an object of `PackageSpecMutator` and apply series of mutations. I will try to produce a simple high-level design for this problem using this design pattern shortly. If the team members feel is not an acceptable pattern, then I am happy to engage further and learn more.

Reply:
If through PackageSpec mutation we understand the fact that a packageSpec is changed after instantiation, before and after this change the  pattern to create a package spec was in a mutable fashion. The steps to create a packageSpec is: 
1. Instantiate a PackageSpec with the default constructor
2. Set properties etc.  At this point the instance is mutated.

This change does not change any of this. 
With this change a PackageSpec creator can construct its PackageSpec dependencies with the data needed. In this case for a project opted in CPVM the Package References versions will be set from CPV.  

There are discussions to make the PackageSpec (and maybe other types) immutable. I think that the link sent it will be a great material for this effort.